### DISCUSSION TITLE
Advanced fan mode

### DISCUSSION THREAD

**Discussion by teackot** [BeardOverflow/msi-ec#15] [Ideas]
Now that we want to implement the fan curve configuration, we need a way to expose it to userspace.

The fan curve consists of 7 fan speed parameters and 6 temperature thresholds.

I have some ideas of how to represent these settings in userspace:

1. Treat the fan curve as 7 settings of 2 values - temperature and speed (except the 1st one which only has speed). Create 7 attributes that accept or return these 2 values
2. Same as 1, but only have one attribute which accepts 3 values: number of one of 7 settings, temp and speed. It will have to return the whole configuration at once when read.
3. Have one attribute that accepts or returns the whole configuration at once. Probably easy to implement.
4. Treat the fan curve as 2 settings: an array of 7 speeds and an array of 6 temp thresholds. Have 2 attribures that accept or return these arrays. Probably the easiest one to implement

**Comment by teackot** [comment]
Kernel prefers simple algorithms so I like the number 4.

Number 1 seems pretty flexible if we can dynamically create attributes without hardcoding all the 14 functions, but I'm not sure if it's worth the effort. It'll also require the msi_platform_probe() to be modified.

**Reply by giga-bytes-dev** [reply]
I thing 4 algorithm correct and true. Software from msi uses the same algorithm.In the application from msi 7 vertical sliders

**Reply by giga-bytes-dev** [reply]
Perhaps you need to look at how it is implemented in other drivers for linux.I'll take a look and reply again

**Reply by teackot** [reply]
I think MSI Center's implementation is closer to 1, because they have sliders with 2 coordinates. But I agree that we should take a look at other drivers' fan curve implementations

**Reply by giga-bytes-dev** [reply]
![image](https://user-images.githubusercontent.com/83925814/224929574-7e64414e-c06a-447a-8c7d-fca6c6f79d4b.png)

**Reply by giga-bytes-dev** [reply]
examples of addresses in https://github.com/YoyPa/isw/issues/269

**Reply by giga-bytes-dev** [reply]
-----------CPU----------- -----------GPU-----------
Value set @ address Value set @ address
0x37(55°C) 0x6a(byte106) 0x37(55°C) 0x82(byte130)
0x3c(60°C) 0x6b(byte107) 0x3c(60°C) 0x83(byte131)
0x46(70°C) 0x6c(byte108) 0x46(70°C) 0x84(byte132)
0x4e(78°C) 0x6d(byte109) 0x50(80°C) 0x85(byte133)
0x55(85°C) 0x6e(byte110) 0x52(82°C) 0x86(byte134)
0x5a(90°C) 0x6f(byte111) 0x57(87°C) 0x87(byte135)

0x0(0%) 0x72(byte114) 0x0(0%) 0x8a(byte138)
0x2d(45%) 0x73(byte115) 0x37(55%) 0x8b(byte139)
0x3c(60%) 0x74(byte116) 0x41(65%) 0x8c(byte140)
0x46(70%) 0x75(byte117) 0x46(70%) 0x8d(byte141)
0x4b(75%) 0x76(byte118) 0x4b(75%) 0x8e(byte142)
0x50(80%) 0x77(byte119) 0x50(80%) 0x8f(byte143)
0x50(80%) 0x78(byte120) 0x50(80%) 0x90(byte144)

**Comment by glpnk** [comment]
[Liquidctl](https://github.com/liquidctl/liquidctl/tree/main) for `nzxt-kraken3` use pairs of `temp pwm` for each point (TODO: proof)
>temp[1-2]_auto_point[1-40]_pwm

+ https://docs.kernel.org/hwmon/nzxt-kraken3.html
+ https://github.com/liquidctl/liquidtux/blob/master/drivers/hwmon/nzxt-kraken3.c

Currently unclear for me:
+ each point need to be in percents or mapped to 0-255 for compatibility (like in `nzxt-kraken3` driver)

**Comment by glpnk** [comment]
hwmon API guide https://docs.kernel.org/hwmon/sysfs-interface.html
https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface

That's what's we need

pwm[1-*]_auto_point[1-*]_pwm / pwm[1-*]_auto_point[1-*]_temp / pwm[1-*]_auto_point[1-*]_temp_hyst

    Define the PWM vs temperature curve.

temp[1-*]_auto_point[1-*]_pwm / temp[1-*]_auto_point[1-*]_temp / temp[1-*]_auto_point[1-*]_temp_hyst

    Define the PWM vs temperature curve.

> There is a third case where trip points are associated to both PWM output channels and temperature channels: the PWM values are associated to PWM output channels while the temperature values are associated to temperature channels. In that case, the result is determined by the mapping between temperature inputs and PWM outputs. When several temperature inputs are mapped to a given PWM output, this leads to several candidate PWM values. The actual result is up to the chip, but in general the highest candidate value (fastest fan speed) wins.


## Other way in [amdgpu driver](https://docs.kernel.org/gpu/amdgpu/thermal.html)


### fan_curve

The amdgpu driver provides a sysfs API for checking and adjusting the fan control curve line.

Reading back the file shows you the current settings(temperature in Celsius degree and fan speed in pwm) applied to every anchor point of the curve line and their permitted ranges if changable.

Writing a desired string(with the format like “anchor_point_index temperature fan_speed_in_pwm”) to the file, change the settings for the specific anchor point accordingly.

When you have finished the editing, write “c” (commit) to the file to commit your changes.

If you want to reset to the default value, write “r” (reset) to the file to reset them

There are two fan control modes supported: auto and manual. With auto mode, PMFW handles the fan speed control(how fan speed reacts to ASIC temperature). While with manual mode, users can set their own fan curve line as what described here. Normally the ASIC is booted up with auto mode. Any settings via this interface will switch the fan control to manual mode implicitly.

**Reply by glpnk** [reply]
Basic fan mode partial explanation #131

https://asset.msi.com/global/picture/faq/10011432@2016-0322-0318-069981@faq_0000000002155_en.pdf

**Reply by glpnk** [reply]
https://github.com/YoyPa/isw/issues/204#issuecomment-904012572
